panic: runtime error: invalid memory address or nil pointer dereference

Java Spring Boot has the ability to cache objects in hazelcast.

Those objects can not be read with go/hazelcast:


```
panic: runtime error: invalid memory address or nil pointer dereference                                                                   
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x50cccb]

goroutine 1 [running]:
local/isocial/boof/vendor/github.com/hazelcast/hazelcast-go-client/internal/serialization.(*Service).ToObject(0xc4201982a
0, 0xc4201a8940, 0x89b0a0, 0xc42023ec50, 0x0, 0x0)
        /home/roidelapluie/go/src/local/isocial/boof/vendor/github.com/hazelcast/hazelcast-go-client/internal/serializati
on/serialization.go:74 +0x14b
local/isocial/boof/vendor/github.com/hazelcast/hazelcast-go-client/internal/colutil.DataToObjectPairCollection(0xc4201b60
08, 0x1, 0x1, 0xc4201982a0, 0x5d4b94, 0xc420248a80, 0xc420240930, 0x0, 0x0)
        /home/roidelapluie/go/src/local/isocial/boof/vendor/github.com/hazelcast/hazelcast-go-client/internal/colutil/uti
l.go:64 +0x124
local/isocial/boof/vendor/github.com/hazelcast/hazelcast-go-client/internal.(*proxy).decodeToPairSliceAndError(0xc4202021
20, 0xc420240930, 0x0, 0x0, 0x9871e0, 0x412318, 0x10, 0x8ba740, 0x1, 0xc42023ec00)
        /home/roidelapluie/go/src/local/isocial/boof/vendor/github.com/hazelcast/hazelcast-go-client/internal/proxy.go:22
5 +0x94
local/isocial/boof/vendor/github.com/hazelcast/hazelcast-go-client/internal.(*mapProxy).EntrySet(0xc4201cc018, 0xc4201769
60, 0xc420022f00, 0x11, 0xcecbc0, 0x0)
        /home/roidelapluie/go/src/local/isocial/boof/vendor/github.com/hazelcast/hazelcast-go-client/internal/map.go:408
+0x95
main.fillMaps(0xc420026ed0, 0xe, 0xc420026ef0, 0xe, 0xc420022e80, 0x20, 0xc42016c4c0, 0x1, 0x1, 0x9d7340, ...)
        /home/roidelapluie/go/src/local/isocial/boof/boof.go:57 +0x736
main.run(0xc420026ed0, 0xe, 0xc420026ef0, 0xe, 0xc420022e80, 0x20, 0xc42016c4c0, 0x1, 0x1)
        /home/roidelapluie/go/src/local/isocial/boof/boof.go:27 +0x1cb
main.main()
        /home/roidelapluie/go/src/local/isocial/boof/main.go:33
```

rev 204d50f544905524548dfac0718134e836d5fbcf

--- a/internal/serialization/serialization.go
+++ b/internal/serialization/serialization.go
@@ -66,10 +66,14 @@ func (s *Service) ToObject(data *Data) (interface{}, error) {
 	if data == nil {
 		return nil, nil
 	}
-	if data.GetType() == 0 {
+	typeID := data.GetType()
+	if typeID == 0 {
 		return data, nil
 	}
-	var serializer = s.registry[data.GetType()]
+	serializer, ok := s.registry[typeID]
+	if !ok {
+		return nil, core.NewHazelcastSerializationError(fmt.Sprintf("there is no suitable de-serializer for type %d", typeID), nil)
+	}
 	dataInput := NewObjectDataInput(data.Buffer(), DataOffset, s, s.serializationConfig.IsBigEndian())
 	return serializer.Read(dataInput)
 }

--- a/internal/serialization/serialization_test.go
+++ b/internal/serialization/serialization_test.go
@@ -22,6 +22,7 @@ import (
 
 	"github.com/hazelcast/hazelcast-go-client/config"
 	"github.com/hazelcast/hazelcast-go-client/serialization"
+	"github.com/hazelcast/hazelcast-go-client/test/assert"
 )
 
 const (
@@ -284,4 +285,15 @@ func TestSerializeData(t *testing.T) {
 	if !reflect.DeepEqual(data, serializedData) {
 		t.Error("data type should not be serialized")
 	}
+}
+
+func TestUndefinedDataDeserialization(t *testing.T) {
+	s, _ := NewSerializationService(config.NewSerializationConfig())
+	dataOutput := NewPositionalObjectDataOutput(1, s, s.serializationConfig.IsBigEndian())
+	dataOutput.WriteInt32(0) // partition
+	dataOutput.WriteInt32(-100)
+	dataOutput.WriteUTF("Furkan")
+	data := &Data{dataOutput.buffer}
+	_, err := s.ToObject(data)
+	assert.ErrorNotNil(t, err, "err should not be nil")
 }

